pkg/controller/plan/adapter/openstack/builder.go

```diff
@@ -823,23 +824,92 @@ func (r *Builder) TemplateLabels(vmRef ref.Ref) (labels map[string]string, err e
 
 	if os != UnknownOS && version != "" {
 		os = fmt.Sprintf("%s%s", os, version)
+		if distro == CentOS && len(version) >= 1 && (version[:1] == "8" || version[:1] == "9") {
+			os = fmt.Sprintf("%s.stream%s", distro, version)
+		} else if os == Windows {
+			switch {
+			case strings.Contains(version, "2k12") || strings.Contains(version, "2012"):
+				os = fmt.Sprintf("%s.2k12", os)
+			case strings.Contains(version, "2k16") || strings.Contains(version, "2016"):
+				os = fmt.Sprintf("%s.2k16", os)
+			case strings.Contains(version, "2k19") || strings.Contains(version, "2019"):
+				os = fmt.Sprintf("%s.2k19", os)
+			case strings.Contains(version, "2k22") || strings.Contains(version, "2022"):
+				os = fmt.Sprintf("%s.2k22", os)
+			case len(version) >= 2 && (version[:2] == "10" || version[:2] == "11"):
+				os = fmt.Sprintf("%s.%s", os, version)
+			default:
+				os = DefaultWindows
+			}
+		}
+	}
+	return
+}
+
+func (r *Builder) getTemplateOs(vm *model.Workload) (os string) {
+	distro := ""
+	version := ""
+
+	if osDistro, ok := vm.Image.Properties[OsDistro]; ok {
+		distro = osDistro.(string)
+	}
+
+	if osVersion, ok := vm.Image.Properties[OsVersion]; ok {
+		version = osVersion.(string)
+	}
+
+	switch distro {
+	case ArchLinux, Debian, Gentoo, Mandrake, Mandriva, MES:
+		os = UnknownOS
+	case FreeBSD, OpenBSD, NetBSD:
+		os = UnknownOS
+	case RHEL, CentOS, Fedora, Ubuntu, OpenSUSE, Windows:
+		os = distro
+	case SLED:
+		os = OpenSUSE
+	case MSDOS:
+		os = Windows
+	default:
+		os = UnknownOS
+	}
```

Comment:
it's better after the split but there's still room for reducing code duplication